vxquery-core/src/main/java/org/apache/vxquery/hdfs2/HDFSFunctions.java

```diff
@@ -0,0 +1,463 @@
+                }
+            }
+        } catch (ParserConfigurationException e) {
+            System.err.println(e);
+        } catch (SAXException e) {
+            System.err.println(e);
+        } catch (IOException e) {
+            System.err.println(e);
+        }
+    }
+
+    /**
+     * Writes the schedule to a temporary file, then uploads the file to the HDFS.
+     */
+    public void addScheduleToDistributedCache() {
+        String filepath = "/tmp/splits_schedule.txt";
+        String dfs_path = "vxquery_splits_schedule.txt";
+        PrintWriter writer;
+        try {
+            writer = new PrintWriter(filepath, "UTF-8");
+            for (int split : this.schedule.keySet()) {
+                writer.write(split + "," + this.schedule.get(split));
+            }
+            writer.close();
+        } catch (FileNotFoundException e) {
+            System.err.println(e);
+        } catch (UnsupportedEncodingException e) {
+            System.err.println(e);
+        }
+        // Add file to HDFS
+        this.put(filepath, dfs_path);
+    }
+
+    public RecordReader getReader() {
+
+        List<FileSplit> fileSplits = new ArrayList<FileSplit>();
+        for (int i = 0; i < splits.size(); i++) {
+            fileSplits.add((FileSplit) splits.get(i));
+        }
+        FileSplitsFactory splitsFactory;
+        try {
+            splitsFactory = new FileSplitsFactory(fileSplits);
+            List<FileSplit> inputSplits = splitsFactory.getSplits();
+            ContextFactory ctxFactory = new ContextFactory();
+            int size = inputSplits.size();
+            for (int i = 0; i < size; i++) {
+                /**
+                 * read the split
+                 */
+                TaskAttemptContext context;
+                try {
+                    context = ctxFactory.createContext(job.getConfiguration(), i);
+                    RecordReader reader = inputFormat.createRecordReader(inputSplits.get(i), context);
+                    reader.initialize(inputSplits.get(i), context);
+                    return reader;
+                } catch (HyracksDataException e) {
+                    System.err.println(e);
+                } catch (IOException e) {
+                    System.err.println(e);
+                } catch (InterruptedException e) {
+                    System.err.println(e);
+                }
+            }
+        } catch (HyracksDataException e1) {
+            // TODO Auto-generated catch block
+            e1.printStackTrace();
+        }
+        return null;
+    }
+
+    /**
+     * @return schedule.
+     */
+    public HashMap<Integer, String> getSchedule() {
+        return this.schedule;
+    }
+
+    /**
+     * Return the splits belonging to this node for the existing schedule.
+     * 
+     * @param node
+     * @return
+     */
+    public ArrayList<Integer> getScheduleForNode(String node) {
+        ArrayList<Integer> node_schedule = new ArrayList<Integer>();
```

Comment:
side question: Does this scheduler make the local cpu process the local data? Please post a comment to the mailing list in the cluster diagrams thread.